pytorch_lightning/callbacks/swa.py

```diff
@@ -0,0 +1,206 @@
+
+            annealing_strategy (str): "cos" or "linear"; specifies the annealing
+                strategy: "cos" for cosine annealing, "linear" for linear annealing
+                (default: "cos")
+
+            device (torch.device, optional): if provided, the averaged model will be
+                stored on the `device`. Default: `cpu`
+
+        """
+
+        if not isinstance(swa_epoch_start, int) or isinstance(swa_epoch_start, int) and swa_epoch_start < 0:
+            raise MisconfigurationException("swa_epoch_start should be positive integer.")
+
+        if not isinstance(swa_lrs, (float, list)) \
+           or isinstance(swa_lrs, float) and swa_lrs <= 0 \
+           or isinstance(swa_lrs, list) and not all(lr > 0 for lr in swa_lrs):
+            raise MisconfigurationException("swa_lrs should be a positive float or a list of positive float.")
+
+        if isinstance(avg_fn, Callable):
+            raise MisconfigurationException("avg_fn should be function.")
+
+        self._swa_start = swa_epoch_start
+        self._swa_lrs = swa_lrs
+        self._annealing_epochs = annealing_epochs
+        self._annealing_strategy = annealing_strategy
+        self._avg_fn = avg_fn
+        self._device = device
+        self._model_contains_batch_norm = None
+
+    @property
+    def swa_model(self):
+        return getattr(self, "_average_model")
+
+    @staticmethod
+    def pl_module_contains_batch_norm(pl_module):
+        for module in pl_module.modules():
+            if isinstance(module, nn.modules.batchnorm._BatchNorm):
+                return True
+        return False
+
+    def reset_batch_norm_and_save_state(self, average_model, device):
+        self.momenta = {}
+        for module in average_model.modules():
+            if isinstance(module, nn.modules.batchnorm._BatchNorm):
+                running_mean_dtype = module.running_mean.dtype
+                running_var_dype = module.running_var.dtype
+                module.running_mean = torch.zeros_like(module.running_mean, device=device, dtype=running_mean_dtype)
+                module.running_var = torch.ones_like(module.running_var, device=device, dtype=running_var_dype)
+                self.momenta[module] = module.momentum
+                module.momentum = None
+                module.num_batches_tracked *= 0
+
+    def apply_momemta(self):
+        for bn_module in self.momenta.keys():
+            bn_module.momentum = self.momenta[bn_module]
+
+    def on_fit_start(self, trainer, pl_module):
+        self._average_model = LightningAveragedModel(pl_module, device=self._device, avg_fn=self._avg_fn)
+        optimizers = trainer.optimizers
+        lr_schedulers = trainer.lr_schedulers
+
+        if len(optimizers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 optimizer.")
+
+        if len(lr_schedulers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 lr_scheduler.")
```

Comment:
we should probably state this with a note in documentation

Reply:
Done !